http-server/src/main/java/org/triplea/server/lobby/game/listing/GameListing.java

```diff
@@ -0,0 +1,204 @@
+package org.triplea.server.lobby.game.listing;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.triplea.http.client.lobby.game.listing.LobbyGame;
+import org.triplea.http.client.lobby.game.listing.LobbyGameListing;
+import org.triplea.lobby.server.db.dao.ModeratorAuditHistoryDao;
+
+/**
+ * Class that stores the set of games in the lobby. Games are identified by a combination of two
+ * values, the api-key of the user that created the game and a UUID assigned when the game is
+ * posted. Keep in mind that the 'gameId' is a publicly known value. Therefore we use API key, for
+ * example, to ensure that other users can't invoke endpoints directly to remove the games of other
+ * players.<br>
+ *
+ * <h2>Keep-Alive</h2>
+ *
+ * The game listing has a concept of 'keep-alive' where games, after posting, need to send
+ * 'keep-alive' messages periodically or they will be de-listed. If a game has been de-listed, and
+ * we get a 'keep-alive' message, the client will get a 'false' value back indicating they would
+ * need to (automatically) re-post their game. This way for example if the lobby were to be
+ * restarted, clients sending keep-alives would notice that their game is no longer listed and would
+ * automatically send a game post message to re-post their game.
+ *
+ * <h2>Moderator Boot Game</h2>
+ *
+ * The moderator boot is similar to remove game but there is no check for an API key, any moderator
+ * can boot any game.
+ *
+ * <h2>Event Updates</h2>
+ *
+ * A listener for game created or updated events and another listener game removed are injected into
+ * this class. Those listeners are invoked for the respective events. The listener is then intended
+ * to be a websocket to notify players of the change event.
+ */
+@Builder
+@Slf4j
+// TODO: Project#12 Create a thread that will invoke game reaper periodically to prune dead games
+class GameListing {
+  @NonNull private final Consumer<LobbyGameListing> gameUpdateListener;
+  @NonNull private final Consumer<String> gameRemoveListener;
+  @NonNull private final GameReaper gameReaper;
+  @NonNull private final ModeratorAuditHistoryDao auditHistoryDao;
+
+  private final Map<GameId, LobbyGame> games = new HashMap<>();
+
+  @Builder
+  @EqualsAndHashCode
+  @Getter
+  static class GameId {
+    @NonNull private final String apiKey;
+    @NonNull private final String id;
+  }
+
+  @VisibleForTesting
+  class GameNotFound extends RuntimeException {
```

Comment:
For these custom exceptions. would it be better to suffix them with 'Error' or 'Exception'?

Reply:
Exception maybe, but IMO it's kingdom of nouns situation, it's already clear they are exceptions. An example usage:
```
           .orElseThrow(() -> new GameNotFound(gameId));
```
I don't think that needs an 'Exception' suffix to be more clear.

Reply:
I guess I prefer having the suffix as its probably more consistent with the rest of the exceptions and makes it easier to read. I think some of these like "GameNotFound" could be say an entry in a game list or a response object to a user that clicked on a game that didn't exist. If you prefer to leave as is I'm ok with that. I'll leave it up to you.

Reply:
Principle of least surprise does apply; then again, it's a language level problem that all exceptions are suffixed Exception and the language probably should never have done that. Fighting the language does violate principle of least surprise, I think this where people start to not like Java as a language (particularly bad in J2EE and previous versions, lots of harmful ceremony built into the language).

I'm inclined to see how this plays out, the exception is pretty local and always is surrounded in context for now.